core/org.openjdk.jmc.flightrecorder.rules.jdk/src/main/java/org/openjdk/jmc/flightrecorder/rules/jdk/util/DefaultIItemResultSet.java

```diff
@@ -96,20 +105,36 @@ private void calculateData(IItemCollection input) {
 			IAggregator<?, ?> aggregator = Aggregators.distinct(query.getGroupBy());
 			Set<?> aggregate = input.getAggregate((IAggregator<Set<?>, ?>) aggregator);
 			if (aggregate != null) {
-				for (Object o : aggregate) {
-					IItemCollection rowCollection = input.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
-					Object[] row = newRow();
-					int column = 0;
-					for (; column < attributes.size(); column++) {
-						// Optimization - it is too expensive to do aggregation for these. You simply
-						// get first non-null matching attribute - we're only using this for the group by today.
-						row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+				ExecutorService exec = Executors.newWorkStealingPool();
+				try {
+					for (final Object o : aggregate) {
+						exec.submit(new Runnable() {
+							@Override
+							public void run() {
+								IItemCollection rowCollection = newInput
+										.apply(ItemFilters.equals((IAttribute) query.getGroupBy(), o));
+								Object[] row = newRow();
+								int column = 0;
+								for (; column < attributes.size(); column++) {
+									// Optimization - it is too expensive to do aggregation for these. You simply
+									// get first non-null matching attribute - we're only using this for the group by today.
+									row[column] = getFirstNonNull(rowCollection, attributes.get(column));
+								}
+								for (int j = 0; j < aggregators.size(); j++) {
+									row[column + j] = rowCollection.getAggregate(aggregators.get(j));
+								}
+								synchronized (data) {
+									data.add(row);
+								}
+							}
+						});
 					}
-					for (int j = 0; j < aggregators.size(); j++) {
-						row[column + j] = rowCollection.getAggregate(aggregators.get(j));
-					}
-					data.add(row);
+				} finally {
+					exec.shutdown();
+					exec.awaitTermination(ClassLeakingRule.CONFIGURED_TIMEOUT, TimeUnit.MINUTES);
 				}
```

Comment:
This timeout should be passed as a parameter instead as this could differ if rules are run concurrently with different configurations for this. Also if we would want to use this method from another rule it would be impossible to set the timeout properly.